core/drivers/imx_i2c.c

```diff
@@ -0,0 +1,397 @@
+
+	return TEE_SUCCESS;
+}
+
+static TEE_Result i2c_read_byte(uint8_t bid, uint8_t *p)
+{
+	TEE_Result ret = TEE_SUCCESS;
+
+	*p = i2c_io_read8(bid, I2DR);
+	ret = i2c_sync_bus(bid, &isr_active, NULL);
+	i2c_io_write8(bid, I2SR, 0);
+
+	return ret;
+}
+
+static TEE_Result i2c_write_data(uint8_t bid, const uint8_t *buf, int len)
+{
+	TEE_Result ret = TEE_SUCCESS;
+	uint32_t tmp = 0;
+
+	if (!len)
+		return TEE_SUCCESS;
+
+	tmp = i2c_io_read8(bid, I2CR) | I2CR_MTX | I2CR_TX_NO_AK;
+	i2c_io_write8(bid, I2CR, tmp);
+
+	while (len--) {
+		ret = i2c_write_byte(bid, *buf++);
+		if (ret != TEE_SUCCESS)
+			return ret;
+	}
+
+	return ret;
+}
+
+static TEE_Result i2c_read_data(uint8_t bid, uint8_t *buf, int len)
+{
+	TEE_Result ret = TEE_SUCCESS;
+	uint8_t dummy = 0;
+	uint32_t tmp = 0;
+
+	if (!len)
+		return TEE_SUCCESS;
+
+	tmp = i2c_io_read8(bid, I2CR) & ~I2CR_MTX;
+	tmp = (len == 1) ? tmp | I2CR_TX_NO_AK : tmp & ~I2CR_TX_NO_AK;
+	i2c_io_write8(bid, I2CR, tmp);
+	i2c_io_read8(bid, I2DR);
+
+	ret = i2c_read_byte(bid, &dummy);
+	if (ret != TEE_SUCCESS)
+		return ret;
+
+	/*
+	 * i.MX 8M Mini APRM Rev2 08/2019:
+	 * a data transfer ends when the master signals a stop; for a master
+	 * receiver to terminate a transfer it must inform the slave transmiter
+	 * by not acknowledging the last data byte. This is done by setting the
+	 * transmit acknowledge bit before reading the next-to-last byte.
+	 */
+	do {
+		if (len == 2) {
+			tmp = i2c_io_read8(bid, I2CR) | I2CR_TX_NO_AK;
+			i2c_io_write8(bid, I2CR, tmp);
+		}
+
+		ret = i2c_read_byte(bid, buf++);
+		if (ret != TEE_SUCCESS)
+			return ret;
+	} while (len--);
+
+	return ret;
+}
+
+/* i.MX 8M Mini APRM Rev2 08/2019: steps previous to a read/write operation  */
+static TEE_Result i2c_init_transfer(uint8_t bid, uint8_t chip)
+{
+	TEE_Result ret = TEE_SUCCESS;
+	uint32_t tmp = 0;
+
+	ret = i2c_idle_bus(bid);
+	if (ret != TEE_SUCCESS)
+		return ret;
+
+	/* enable the interface */
+	i2c_io_write8(bid, I2CR, I2CR_IEN);
+
+	/* 16.1.4.1 start master transaction */
+	tmp = i2c_io_read8(bid, I2CR) | I2CR_MSTA;
+	i2c_io_write8(bid, I2CR, tmp);
+
+	/* 16.1.4.2. on multimaster bus system, wait unil we have the bus */
+	ret = i2c_sync_bus(bid, &bus_is_busy, NULL);
+	if (ret != TEE_SUCCESS)
+		return ret;
+
+	/* put the slave address on the bus */
+	return i2c_write_data(bid, &chip, 1);
+}
+
+TEE_Result imx_i2c_read(uint8_t bid, uint8_t chip, uint8_t *buf, int len)
+{
+	TEE_Result ret = TEE_SUCCESS;
+
+	if (bid > ARRAY_SIZE(i2c_bus) - 1)
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	if ((len && !buf) || chip > 0x7F)
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	ret = i2c_init_transfer(bid, chip << 1 | 1);
+	if (ret == TEE_SUCCESS)
+		ret = i2c_read_data(bid, buf, len);
+
+	if (i2c_idle_bus(bid) != TEE_SUCCESS)
+		IMSG("bus not idle");
+
+	return ret;
+}
+
+TEE_Result imx_i2c_write(uint8_t bid, uint8_t chip, const uint8_t *buf, int len)
+{
+	TEE_Result ret = TEE_SUCCESS;
+
+	if (bid > ARRAY_SIZE(i2c_bus) - 1)
```

Comment:
minor suggestion: `if (bid >= ARRAY_SIZE(i2c_bus))` here and and at lines 298, 336, 349.

Reply:
yes